### PR TITLE
Removing sort_string and id from research datums.

### PR DESCRIPTION
**Pull request by MistakeNot4892**
Breaking changes from #116 out into a separate PR.

- Removed `sort_string` and made recipe sorting use `name`.
- Removed `id` and made the R&D interfaces use a reference to the datum instead.

### CODE REVIEW

**Comment by MistakeNot4892** [comment]
Ready to go.

File: code/modules/research/research.dm [hunk 1]
```diff
@@ -91,14 +91,14 @@ research holder datum.
...
 
 /datum/research/proc/AddDesign2Known(var/datum/design/D)
-	if(!known_designs.len) // Special case
+	if(!length(known_designs))
 		known_designs.Add(D)
 		return
 	for(var/i = 1 to known_designs.len)
 		var/datum/design/A = known_designs[i]
-		if(A.id == D.id) // We are guaranteed to reach this if the ids are the same, because sort_string will also be the same
+		if(A.name == D.name)
```
Comment (PsiOmegaDelta) [inline code/modules/research/research.dm]: Wouldn't `A.type == D.type` work here?

File: code/modules/research/research.dm [hunk 2]
```diff
@@ -91,14 +91,14 @@ research holder datum.
...
 			return
-		if(A.sort_string > D.sort_string)
+		if("[A.name]" > "[D.name]")
```
Comment (PsiOmegaDelta) [inline code/modules/research/research.dm]: Doesn't `if(A.name > D.name)` work?
Seems a little odd to create a new string out of what should already be a string

Comment (PsiOmegaDelta) [inline code/modules/research/research.dm]: Feels like this should have the same functionality
```
/datum/research/proc/AddDesign2Known(var/datum/design/D)
    if(D.name in known_designs)
        return
    ADD_SORTED(known_designs, D.name, /proc/cmp_text_asc)
    known_designs[D.name] = D
```

Comment (MistakeNot4892) [reply code/modules/research/research.dm]: It runtimes on a null - I actually had name initialize and forgot to reverse this.

**Review by PsiOmegaDelta — APPROVED** [review]

**Pull request merged**